### ISSUE TITLE
Top Level Input with binding Equation

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#303]
**Reported by jens&#46;frenkel on 20 Nov 2012 08:18 UTC**
Consider the following Model (Spec 3.3, p. 44)
```
model M
  input Real u;
  input Real u2=2;
end M;
```

If that model is simulated, I would expect u to have a value of 0 and u2 to be 2. However, the latest Dymola implementation says that both u and u2 are 0.

Is this according to the standard?

Note:

If you use model M within a model N
```
model N
  M m(u = 1);
end N;
```
then u = 1 and u2 = 2 as expected.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/906

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 20 Nov 2012 08:26 UTC**
I think this belongs into the Dymola bugtracker.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 Nov 2012 08:37 UTC**
This is most likely not precisely specified in Modelica.

For me the best semantics is that model M as top level to be simulated is wrong and translation should result in an error. Reasons:

 * As a top level model, "input" defines a causal input from the outside to the Modelica model. Therefore "u2" is defined from the outside
 * The statement "u2=2" specifies a default equation for u2. Since there is no modification of M, the default equation holds. 

The two statements are in contradiction and therefore this model should result in an error.

Another semantic definition could be that providing an input from the outside is treated as a modification of model M and therefore the default equation u2=2 is replaced by the default equation provided from the outside (u2=0). This seems to be the semantics that Dymola is using.

In any case, the model is confusing for an end-user (providing a default equation for a variable declared as input).

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 20 Nov 2012 08:51 UTC**
A more intuitive approach could be to regard an input similar to a connector. 
In case of model M, the input u is "not connected", i.e. it has no binding equation and therefore u=0 is automatically generated (maybe a warning should be issued). 
In case of u2 there exists a (default) binding equation, i.e. it is connected, and thus that binding equation is being used.
In case of Model N, a binding equation for u has to be supplied for m to be locally balanced.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 Nov 2012 09:49 UTC**
Replying to [comment:3 anonymous]:


> A more intuitive approach could be to regard an input similar to a connector. 
> In case of model M, the input u is "not connected", i.e. it has no binding equation and therefore u=0 is automatically generated (maybe a warning should be issued). 
> In case of u2 there exists a (default) binding equation, i.e. it is connected, and thus that binding equation is being used.
> In case of Model N, a binding equation for u has to be supplied for m to be locally balanced.

A Modelica model on top level with inputs and outputs is seen as an input/output block. For example, this model can be exported as FMU and has then input and output signals that are set from the environment where the FMU is called. Furthermore, it is possible to linearize this model (whether or not exported as FMU) and the top level inputs/outputs are the inputs/outputs of the linearized model. In Dymola (and I hope in other programs as well), a top level input can be also set from the outside. If no input is given from the outside, then a default of zero is used.

To summarize, a simulation model (top level Modelica model) must be treated in a different way in this respect as a standard Modelica model. I think this is partially defined in the specification but the text must eventually be improved.

**Comment by modelica-trac-importer** [comment]
**Comment by frenkel&#x2e;jens on 20 Nov 2012 14:02 UTC**
Maybe it is more userfriendly to allow the definition of other default values than 0 for inputs. In our example for u2 it would be 2. Then it is possible to set not all inputs but use the block with meaningfull default values.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 20 Dec 2012 13:47 UTC**
Housekeeping, defaulting all new tickets to [milesstone:\*undecided\*] (is normally done automatically)

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 30 Mar 2016 17:00 UTC**
Replying to [comment:4 otter]:
> A Modelica model on top level with inputs and outputs is seen as an input/output block. For example, this model can be exported as FMU and has then input and output signals that are set from the environment where the FMU is called. Furthermore, it is possible to linearize this model (whether or not exported as FMU) and the top level inputs/outputs are the inputs/outputs of the linearized model. In Dymola (and I hope in other programs as well), a top level input can be also set from the outside. If no input is given from the outside, then a default of zero is used.

(Just to be clear - Dymola has been improved to use better defaults in this case.)

> To summarize, a simulation model (top level Modelica model) must be treated in a different way in this respect as a standard Modelica model. I think this is partially defined in the specification but the text must eventually be improved.

The handling of models with external inputs are in fact not described by the specification (as far as I can see) - it is in particular missing from the flat representation. It is also problematic to handle that case since inputs may be differentiated.

I thus propose to close it as expired.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Jun 2016 12:59 UTC**
Housekeeping to make it easier to find the active tickets for the specification.

These tickets are all expired, i.e. old with no action - and not clear that any action is needed. In most cases they are enhancements/discussions without any clear result.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 26 Aug 2016 06:21 UTC**
House keeping.
